Improved EnergyDispersion2D.get_response and associated tests

registerrier opened the pull request:
This introduces a different get_response() where integration is done with a given step in migration rather than with a fixed oversampling of e_reco bins.

- The default value for the migra_step allows a significant decrease of computation time and allow for absolute precision <0.05.
- Tests have been modified to test for a gaussian edisp where we can compute the desired value with satisfactory precision (<5% absolute)
- Note that all tests on EnergyDispersion2D could be done with gaussian edisp rather than stored one.

joleroi commented on gammapy/irf/energy_dispersion.py:
```diff
@@ -731,58 +733,33 @@ def get_response(self, offset, e_true, e_reco=None):
+
+        # Determine positions (bin indices) of e_reco bounds in migration array
+        # We ensure that no negative values are found
+        pos_mig = np.maximum(np.digitize(migra_e_reco, mig_array)-1 , 0)
```
This line is quite involved I think. Could you split it up in 2? Like you already did with the comments

```
# Determine positions (bin indices) of e_reco bounds in migration array
pos_mig_temp = np.digitize(migra_e_reco, mig_array)-1

# We ensure that no negative values are found
pos_mig = np.maximum(pos_mig_temp , 0)
```

joleroi commented on gammapy/irf/tests/test_energy_dispersion.py:
```diff
@@ -93,30 +93,57 @@ def test_evaluation(self):
-        assert_equal(actual, desired)
+        # Here we test get_response with an expected gaussian shape for edisp
+
+        # First we define a fake EnergyDispersion2D
```
The generation of the fake edisp is very nice! However, it should not happen in the test but be available as method on ``EnergyDispersion2D``, similar to [EnergyDispersion.from_gauss](https://github.com/gammapy/gammapy/blob/master/gammapy/irf/energy_dispersion.py#L97) . The test should then just call this function.

joleroi replied:
> Note that all tests on EnergyDispersion2D could be done with gaussian edisp rather than stored one.

This would be also easier if the method you coded would live on the class

registerrier commented:
I have added the _from_gauss function in the EnergyDispersion2D class. 
I corrected a small issue that makes agreement between analytical solution and get_response better.
Use step of 1e-2 in the test gives result <0.05 in absolute error.

cdeil commented on gammapy/irf/energy_dispersion.py:
```diff
@@ -626,6 +626,48 @@ def offset(self):
+        if offset is None:
+            offset = np.linspace(0.,3.,10)*u.deg
+
+        from scipy.special import erf
```
We always put delayed imports at the very top of methods. Please do it here also for consistency with the rest of Gammapy.

registerrier replied:
It is already imported in from_gauss from EnergyDispersion.
I think if we put it on the top, we have to add @requires_dependency('spicy') for EnergyDispersion test as well.
Do we want that?

cdeil commented on gammapy/irf/energy_dispersion.py:
```diff
@@ -626,6 +626,48 @@ def offset(self):
+             Bin edges of offset
+         """
+        if offset is None:
+            offset = np.linspace(0.,3.,10)*u.deg
```
I know we have defaults in a lot of places that really only work well for HESS, like 3 deg FOV size here.
Probably we should remove those, no?
I.e. the caller must be explicit, but then we have no HESS-specific hardcoded stuff.
You know, be CTA friendly.

registerrier replied:
Well, it really does not matter here since it is uniform in offset anyway. No?

cdeil commented on gammapy/irf/energy_dispersion.py:
```diff
@@ -702,7 +744,7 @@ def to_energy_dispersion(self, offset, e_true=None, e_reco=None):
                                 data=rm)
 
-    def get_response(self, offset, e_true, e_reco=None):
+    def get_response(self, offset, e_true, e_reco=None, mig_step=5e-3):
```
I think we call it `migra` pretty consistenly everywhere (also in the spec, and in Gammapy).
So suggest to change `mig_step` to `migra_step` here.

registerrier replied:
Done

cdeil commented on gammapy/irf/tests/test_energy_dispersion.py:
```diff
@@ -93,30 +93,42 @@ def test_evaluation(self):
-        assert_equal(actual, desired)
+        # Here we test get_response with an expected gaussian shape for edisp
+        from scipy.special import erf
+        # First we define a fake EnergyDispersion2D
```
Suggest to split this whole part creating the test edisp object out into a separate method:
```
@staticmethod
def make_test_edisp2d():
    ...
    return edisp
```
Then if one ever needs to test it or if we want to call it from somewhere else, it's possible.
E.g.
```
from gammapy.irf.tests.test_energy_dispersion import TestEnergyDispersion2D
edisp = TestEnergyDispersion2D.make_test_edisp2d()
# play with it interactively or use it from other tests
```

registerrier replied:
Since we also use mu, sigma, trues and migras array later it is simpler to keep it inside the call and not use an external make function

registerrier commented:
Modified test_data_store to add absolute tolerance of 2% to result.

joleroi commented on gammapy/irf/energy_dispersion.py:
```diff
@@ -626,6 +626,48 @@ def offset(self):
         return self.data.axis('offset')
 
     @classmethod
+    def _from_gaus(cls, e_true, migra, bias, sigma, offset=None):
```
:+1:  for the public method (like you put it). I can be useful e.g. for simulations

joleroi commented:
As far as I can see, we merge this and close https://github.com/gammapy/gammapy/pull/894 ?

joleroi commented:
This looks like it was introduced in this PR
https://travis-ci.org/gammapy/gammapy/builds/204946729
I will have a look
